Implemented autosave on a 5-second timer

Levi-Lesches opened the pull request:
Fixes #2871 by saving the editor content to local storage under they key `user_input` every 5 seconds, and in `dispose`

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

Levi-Lesches commented:
@parlough I fixed the double quotes

parlough commented on pkgs/dartpad_ui/lib/editor/editor.dart:
```diff
@@ -177,10 +177,17 @@ class _EditorWidgetState extends State<EditorWidget> implements EditorService {
 
+  Timer? _autosaveTimer;
+  void _autosave([Timer? timer]) {
+    final content = codeMirror?.getDoc().getValue();
+    if (content == null) return;
```
I think it'd be more consistent to rely on the app model here and help if/when we move away from code mirror.

```suggestion
    final content = widget.appModel.sourceCodeController.text;
```

Levi-Lesches replied:
Fixed, and I added a check for `isEmpty` instead of null

parlough commented on pkgs/dartpad_ui/lib/main.dart:
```diff
@@ -277,7 +278,8 @@ class _DartPadMainPageState extends State<DartPadMainPage>
             flutterSampleId: widget.flutterSampleId,
             channel: widget.initialChannel,
-            fallbackSnippet: Samples.getDefault(type: 'dart'))
+            fallbackSnippet: web.window.localStorage.getItem('user_input') ??
+                Samples.getDefault(type: 'dart'))
```
Do you think it's worth it to convert this fallback to a callback, so we don't unnecessarily retrieve the old snippet?

Levi-Lesches replied:
I think ?? is short circuiting, so it should be fine?

Levi-Lesches replied:
Oh, if you meant not retrieving the old snippet or user input in case a gist id or otherwise is provided. 

I tried making it a callback, and got some very strange errors on the web side. Each time I restarted, it was that JS can't read some property, but it was a different property each time. Once it didn't even let the flutter tool connect because the bootstrapper got stuck. Not sure what's up with that but I'll leave the PR as-is for now

Levi-Lesches replied:
@parlough I Implemented this in the latest set of commits. I think there was something wrong with my Flutter installation last time I was working on this. Just upgraded to 3.27.0 with no issues.

parlough reviewed (approved):
Sweet! Thanks so much for implementing this again :D

Levi-Lesches commented:
@parlough Any updates?

Levi-Lesches commented:
Sorry for the annoying pings, just want to make sure this doesn't slip between the cracks

parlough commented:
@Levi-Lesches Sorry for the delay, I've been away.

The change looks good to me, but before landing, we need to wait for others to determine DartPad's storage policies and how we will inform users of them.

I'll land this PR when able and ping you if any changes are needed.

Thanks again!

Levi-Lesches commented:
Hi @parlough, any updates?

johnpryan commented:
We added a cookie banner now (https://github.com/dart-lang/dart-pad/pull/3097), so we should be OK to use local storage.

Levi-Lesches commented:
I'm trying to add an integration test that tests: 
- entering text really does save content to the local storage
- loading the page results in the local storage text being displayed, _unless_
- if the page is loaded with a Gist, checks that the Gist is loaded instead of the local storage

This will complicate CI a bit, so I'm updating the README with exact instructions (they're essentially copied from the [Flutter docs](https://docs.flutter.dev/testing/integration-tests#test-in-a-web-browser).

Levi-Lesches commented:
I ended up going with unit tests instead as I couldn't get the integration tests to work: `WidgetTester.enterText(String)` calls `WidgetTester.showKeyboard()`, which requires an `EditableText` widget somewhere in the subtree, but the current setup with CodeMirror forwards keyboard input directly to the CodeMirror library without using Flutter's `TextInput`. 

The test cases above are still tested, though, so I'm ready for another review.

johnpryan commented on pkgs/dartpad_ui/lib/editor/editor.dart:
```diff
@@ -303,6 +311,8 @@ class _EditorWidgetState extends State<EditorWidget> implements EditorService {
   void dispose() {
     listener?.cancel();
+    _autosaveTimer?.cancel();
+    _autosave();
```
Why is _autosave() called after the editor is disposed?

Levi-Lesches replied:
As far as I'm aware, the editor itself isn't disposed at this point. The intention here was to do one last auto-save before closing the editor. But I couldn't get it to trigger in my tests so I removed it.

johnpryan commented on pkgs/dartpad_ui/test/autosave_test.dart:
```diff
@@ -0,0 +1,128 @@
+
+void main() {
+  const channel = Channel.stable;
+  group('Autosave:', () {
```
I would take the colon out since it's not doing anything useful. The test package formats the output.

johnpryan commented on pkgs/dartpad_ui/lib/main.dart:
```diff
@@ -272,24 +273,27 @@ class _DartPadMainPageState extends State<DartPadMainPage>
-        appServices.performCompileAndRun();
-      }
-    });
+    _initAppServices();
```
This function is only used once, so I would keep the code inlined here.

Levi-Lesches replied:
Fair. I was editing that section anyway so I did that for readability -- it's flatter, the formatted is a little kinder, and it uses `async/await` instead of `.then` directly. 

I reverted that back to `.then()` though

johnpryan replied:
Why not make `initState` async?

Levi-Lesches replied:
It's a Flutter thing: 

```console
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building Builder:
_DartPadMainPageState.initState() returned a Future.
State.initState() must be a void method without an `async` keyword.
Rather than awaiting on asynchronous work directly inside of initState, call a separate method to do
this work without awaiting it.
```

johnpryan replied:
I forgot about that rule, sorry for the churn. Let's keep it as is for now.

johnpryan commented on pkgs/dartpad_ui/lib/editor/editor.dart:
```diff
@@ -177,10 +178,17 @@ class _EditorWidgetState extends State<EditorWidget> implements EditorService {
   }
 
+  Timer? _autosaveTimer;
+  void _autosave([Timer? timer]) {
```
Why does this take a `timer` parameter, if it's not ever used? Let's take it out unless we need it.

Levi-Lesches replied:
Callbacks passed to `Timer.periodic` must accept a `Timer?` parameter. I renamed it to just `_`, but I kept the type `Timer?` and left a comment to clarify why it's there.

johnpryan commented on pkgs/dartpad_ui/lib/local_storage.dart:
```diff
@@ -0,0 +1,13 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'local_storage/stub.dart'
```
Do we need this library? Why can't we use the [localStorage API from package:web](https://pub.dev/documentation/web/latest/web/StorageAccessHandle/localStorage.html)?

Levi-Lesches replied:
Habits of writing cross-platform code, but of course this is a web-only project. Removed the conditional import and replaced it with the real web implementation. 

I was using `web.window.localStorage` though, instead of the API you linked, `web.document.requestStorageAccess()`. Given that DartPad is often embedded into the docs and codelabs, do I need to switch to that?

johnpryan replied:
No, window.localStorage (from package:web) is the correct API, I sent the wrong link.

Levi-Lesches commented:
@johnpryan 

> Habits of writing cross-platform code, but of course this is a web-only project. Removed the conditional import and replaced it with the real web implementation.

Forgot -- the tests don't work with web-only imports. Try running the CI now

johnpryan commented on pkgs/dartpad_ui/lib/local_storage/stub.dart:
```diff
@@ -0,0 +1,16 @@
+import '../local_storage.dart';
+import '../utils.dart';
+
+class LocalStorageImpl extends LocalStorage {
```
Thanks for writing a stub, glad we are able to keep the unit tests working 👍
